File: src/core/ball.py
from game import Game
from player import Player

class Ball:
    x: float = 0
    y: float = 0
    size: float = 50
    speed_x: float = 0
    speed_y: float = 0
    angle: float = 100

    game: Game

    def __init__(self, game: Game):
        self.game = game

    def move(self):
        self.x += self.speed_x
        self.y += self.speed_y

    def impact_player(self, player: Player):
        self.speed_x *= -1
        self.speed_y = (self.y - player.y + player.height / 2) / self.angle
        if 40 > self.speed_x > 0:
            self.speed_x += 0.6
        elif self.speed_x > -40:
            self.speed_x -= 0.6

    def update_speed(self, player1: Player, player2: Player):
        if self.y + self.size >= self.game.height or self.y <= 0:
            self.speed_y *= -1
        elif self.x <= player1.x + player1.width and self.y + self.size >= player1.y and self.y < player1.y + player1.height:
            self.impact_player(player1)
        elif self.x + self.size >= player2.x and self.y + self.size >= player2.y and self.y < player2.y + player2.height:
            self.impact_player(player2)

File: backend/src/Ball.ts
import { Game, Match } from "./Game";
import Player from "./Player";

export default class Ball {
  static readonly ANGLE = 25;
  static readonly SIZE = 50;

  private readonly game: Game;
  private x: number = 0;
  private y: number = 0;
  speedX: number = 0;
  speedY: number = 0;

  constructor(game: Game) {
    this.game = game;
  }

  get goToLeft() {
    return this.speedX < 0;
  }

  get left() {
    return this.x;
  }

  get right() {
    return this.x + Ball.SIZE;
  }

  get top() {
    return this.y;
  }

  get bottom() {
    return this.y + Ball.SIZE;
  }

  get centerX() {
    return this.x + Ball.SIZE / 2;
  }

  get centerY() {
    return this.y + Ball.SIZE / 2;
  }

  resetPos(match: Match & { player2: Player }) {
    this.x = (Game.WIDTH - Ball.SIZE) / 2;
    this.y = (Game.HEIGHT - Ball.SIZE) / 2;
    this.speedX = match.score1 < match.score2 ? -3 : 3;
    this.speedY = 0;
  }

  move() {
    this.x += this.speedX;
    this.y += this.speedY;
  }

  updateSpeed(player1: Player, player2: Player) {
    if (this.bottom >= Game.HEIGHT || this.top <= 0) this.speedY *= -1;
    else if (this.goToLeft && this.left <= player1.right && this.bottom >= player1.top && this.top < player1.bottom)
      this.impactPlayer(player1);
    else if (!this.goToLeft && this.right >= player2.left && this.bottom >= player2.top && this.top < player2.bottom)
      this.impactPlayer(player2);
  }

  impactPlayer(player: Player) {
    this.speedX *= -1;
    this.speedY = ((this.y + Ball.SIZE / 3) - player.centerY) / Ball.ANGLE;
    if (this.speedX > 0 && this.speedX < 40)
      this.speedX += 0.6;
    else if (this.speedX > -40)
      this.speedX -= 0.6;
  }

  toJSON() {
    return {
      x: this.x,
      y: this.y,
      size: Ball.SIZE
    };
  }
}
